src/annoylib.h

```diff
@@ -1207,6 +1207,8 @@ template<typename S, typename T, typename Distance, typename Random, class Threa
           annoylib_showUpdate("File truncation error\n");
     } else {
       _nodes = realloc(_nodes, _s * new_nodes_size);
+      if (!_nodes)
```

Comment:
So the code previously did not properly handle failure of `realloc()` (l. 1209). When we now throw an exception in case of failure we should consider what happens at the callers. Method `_reallocate_nodes()` is called by methods `_allocate_size()` which is in turn called by these methods:
- `add_item_impl()` - everything if fine here
- `build()` - also everything is fine here (including the stuff going on in `ThreadedBuildPolicy::template build()`
- `_make_tree()` - the first call to `_allocate_size()` isn't safe in case of failure, it locks `threaded_build_policy` but doesn't unlock it when an exception is thrown. Subsequent calls to build or query the index will likely run into some sort of deadlock. So one option would be to unlock manually using a `try/catch`.  The deadlock may also be accepted if the caller (MergeTreeIndexAnnoy) discards the entire AnnoyIndex instance but a bit too implicit for my taste. There is a second call to `_allocate_size()` inside `_make_tree()` which has the same problem and solution.